ml-fundamentals/ml-experimentation.md

```diff
@@ -0,0 +1,147 @@
+2. Environment configuration files if applicable (e.g. `requirements.txt` for `venv`, [environment.yml](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file) for `Conda` or [pyrpoject.toml](https://python-poetry.org/docs/pyproject/) for `Poetry`).
+
+#### Virtual environments benefits
+
+- Productization
+- Collaboration
+- Reproducibility
+
+### Source control and folder/package structure
+
+Applied ML projects often contain source code, notebooks, devops scripts, documentation, scientific resources, datasets and more. We recommend to come up with an agreed folder structure to keep resources tidy. Consider deciding upon a generic folder structure for projects (e.g. which contains the folders `data`, `src`, `docs` and `notebooks`), or adopt popular structures like the [CookieCutter Data Science](https://drivendata.github.io/cookiecutter-data-science/) folder structure.
+
+[Source control](https://github.com/microsoft/code-with-engineering-playbook/tree/master/source-control) should be applied to allow collaboration, versioning, code reviews, traceability and backup. In data science projects, source control should be used for code, and the storing and versioning of other  artifacts (e.g. data, scientific literature) should be decided upon depending on the scenario.
+
+#### Folder structure and source control expected outcomes
+
+- Defined folder structure for all users to use, pushed to the repo.
+- [.gitignore](https://git-scm.com/docs/gitignore) file determining which folders should be synced with `git` and which should be kept locally. For example, [this one](https://github.com/drivendata/cookiecutter-data-science/blob/master/%7B%7B%20cookiecutter.repo_name%20%7D%7D/.gitignore).
+- Determine how notebooks are stored and versioned (e.g. [strip output from Jupyter notebooks](https://github.com/kynan/nbstripout))
+
+#### Source control and folder structure benefits
+
+- Collaboration
+- Reproducibility
+- Code quality
+
+### Experiment tracking
+
+Experiment tracking tools allow data scientists and researchers to keep track of previous experiments for better understanding of the experimentation process and for the reproducibility of experiments or models.
+
+#### Types of experiment tracking frameworks
+
+Experiment tracking frameworks differ by the set of features they provide for collecting experiment metadata, and comparing and analysing experiments. In CSE, we mainly use [MLFlow](https://mlflow.org/) on [Databricks](https://databricks.com/product/managed-mlflow) or [Azure ML Experimentation](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-track-experiments). Note that some experiment tracking frameworks require a deployment, while others are SaaS.
+
+#### Experiment tracking outcomes
+
+1. Decide on an experiment tracking framework
+2. Ensure it is accessible to all users
+3. Document set-up on local environments
```

Comment:
There is huge value to providing guidance on setting up the dev inner loop.

We should create all environments (dev, qa, prod, etc.) from source/IaC so they are consistent.

On of those environments created from that IaC script is my local development resource group/resource that map to my branch so when I'm developing something in a branch I can work in an resoure group indpendently. If I need to blow it away I can do so via IaC scripts.

Like virtual environments, we should document that so the team can get set up quickly/consistently.

Not sure if it belongs up here or closer to virtual environments.

Reply:
Great points. Do you think the IaC scripts guidance and multiple environments setup should belong here or should we discuss this on a new page dealing with MLOps fundamentals, or maybe the ml-proposed-process doc which is coming soon.

Reply:
These are all great point and I think they deserve more discussion. I would lean towards them belonging to MLOps fundamentals, but let's get more thoughts on this.